dpctl/_sycl_queue.pyx

```diff
@@ -259,22 +259,20 @@ cdef class SyclQueue:
         """
         cdef DPCTLSyclContextRef CRef
         cdef DPCTLSyclQueueRef QRef
-        cdef DPCTL_DeviceAndContextPair dev_ctx
 
-        dev_ctx = DPCTLDeviceMgr_GetDeviceAndContextPair(DRef)
-        if (dev_ctx.CRef is NULL) or (dev_ctx.DRef is NULL):
+        CRef = DPCTLDeviceMgr_GetCachedContext(DRef)
+        if (CRef is NULL):
+            DPCTLDevice_Delete(DRef)
```

Comment:
`DRef` is the input argument. We should not be deleting it.

Reply:
If the argument device is a sub-device the look-up will fail. I thought we discussed that it would be better to create a new one instead.

Reply:
`_init_queue_from_DPCTLSyclDeviceRef` is returning an error code if the context look up fails. I assume the logic to handle creation of a context for sub-devices is not yet implemented? Do your upcoming changes to `SyclContext` have that change?

As for freeing `DRef`, in the current code it does not cause an issue. Does it? Are you saying from a design style perspective we should let caller free the object? 

Since, `_init_queue_from_DPCTLSyclDeviceRef` is an internal helper that is invoked only by other internal helpers: `_init_queue_from_filter_string`, `_init_queue_from_device`, and `_init_queue_default`, freeing the `DRef` in `_init_queue_from_DPCTLSyclDeviceRef` when we know queue creation failed is fine IMO. But, I can change the code to free it in the callers on returning from  `_init_queue_from_DPCTLSyclDeviceRef` at the cost of adding a check on the return code. Do note that I have a copy in `_init_queue_from_device`, the only place where deleting the `DRef` can have an external side-effect.

Reply:
> As for freeing `DRef`, in the current code it does not cause an issue. Does it? Are you saying from a design style perspective we should let caller free the object?

I see the issue now. `DRef` will leak when queue creation succeeded. I will fix that.

Reply:
> `_init_queue_from_DPCTLSyclDeviceRef` is returning an error code if the context look up fails. I assume the logic to handle creation of a context for sub-devices is not yet implemented? Do your upcoming changes to `SyclContext` have that change?

Ok #334 has the changes.

Reply:
@oleksandr-pavlyk  in `_init_queue_from_device` we are taking the `DRef` from the `SyclDevice` argument and storing it in the `SyclQueue` if queue creation succeeded. I think we need a `DPCTLDevice_Copy` as I have done here. What happens when the `SyclDevice` is deleted thus deallocating the `DPCTLSyclDeviceRef`?